### PR TITLE
Dev branch

### PR DESCRIPTION
**Pull request by Dpham181**

### CODE REVIEW

File: app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/MainActivityTest.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+
+class MainActivityTest {
+    @get:Rule
+    var activityRule: ActivityTestRule<MainActivity> = ActivityTestRule(MainActivity::class.java)
+
+    @Test
+    fun View_test () {
+
+
```
Comment (mahyarbabaie) [inline app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/MainActivityTest.kt]: nit pick: clean up the empty spaces

File: app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/MainActivityTest.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        // this test for textview app name
+        onView(withId(R.id.textView2))
+            .check(matches(withText("Insertion Sort App")))
+        // gifview test id
+        onView(withId(R.id.gifImageView2)).check(matches(withResourceName("gifImageView2")))
+        SystemClock.sleep(3000);
```
Comment (mahyarbabaie) [inline app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/MainActivityTest.kt]: I strongly recommend not to use hard waits. Is there a reason to use a hard sleep timer at the end of the test?

File: app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/MainActivityTest.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    fun ACTIONBAR_TEST() {
+        // Context of the app under test.
+        onView(withId(R.id.sorting)).perform(click())
+        onView(isRoot()).perform(ViewActions.pressBack());
+
+        onView(withId(R.id.aboutus)).perform(click())
+
+        onView(withId(R.id.records)).perform(click())
+        SystemClock.sleep(3000);
+    }
```
Comment (mahyarbabaie) [inline app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/MainActivityTest.kt:43]: Reference my comment above about being against hard waits

File: app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/sortTest.kt [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    @get:Rule
+    var activityRule: ActivityTestRule<sort> = ActivityTestRule(sort::class.java)
+    val InputwithComma = "5,4,6,4,2,1"
+    val InputwithSpace = "5 8 9 3 2 4"
+    val ErrorSize ="1 2 3 4 5 6 7 8 9"
+    val InvalidInput ="11"
+
+    fun TestValidInput(res:String) {
+
+
```
Comment (mahyarbabaie) [inline app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/sortTest.kt:22]: Nit Pick: clean up empty space

File: app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/sortTest.kt [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        onView(withId(R.id.input))
+            .perform(typeText(res))
+        Espresso.closeSoftKeyboard()
+
+        onView(withId(R.id.add)).perform(click())
+        onView(withId(R.id.input))
+            .perform(clearText())
+        SystemClock.sleep(3000);
+
```
Comment (mahyarbabaie) [inline app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/sortTest.kt]: Hard Waits would prefer if they were not used

File: app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/sortTest.kt [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+
+
+    @Test
+    fun Perform_InputValidTest(){
+
+        TestValidInput(InputwithComma)
+
```
Comment (mahyarbabaie) [inline app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/sortTest.kt]: I would not make modular functions for the tests of this scope. Is there a reason to modulize them?

**Review by mahyarbabaie — COMMENTED** [review]
Overall it is pretty solid. For those unit tests, you need to separate out each function per test and not group them in one specific test. I would have separate functions for each invalid test instead of grouping 2 helper functions into one test.

File: app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/MainActivityTest.kt (hunk 1 above)
Comment (Dpham181) [reply app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/MainActivityTest.kt]: great thanks will update that tmr

File: app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/MainActivityTest.kt (hunk 2 above)
Comment (Dpham181) [reply app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/MainActivityTest.kt]: Tried but It didn't work that why I use too many Systemclock. Those were just for helping to see the result on emulator. It needs to wait for each event. We actually don't need it for testing but I just implemented for u guys to see how it works.

File: app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/sortTest.kt (hunk 6 above)
Comment (Dpham181) [reply app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/sortTest.kt]: there was 2 kinds of input with comma and space. I dint want to repeat the event action that why I nested them into one. Is it really a bad ideas on testing ?

Comment (mahyarbabaie) [reply app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/sortTest.kt]: For isolating test failures yes it is a bad idea combining multiple scenarios in one function.

File: app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/MainActivityTest.kt (hunk 2 above)
Comment (mahyarbabaie) [reply app/src/androidTest/java/edu/fullerton/AdvSofwareProcess/InsertionSortApp/MainActivityTest.kt]: We dont need to see the result on the emulator so please remove all of the hard waits. We do not want to slow down automated tests. Idea is to make them fast

**Review by mahyarbabaie — APPROVED** [review]

**Pull request merged**